### PR TITLE
#162170716 Implement favorite/unfavorite article functionality

### PR DESCRIPTION
**Pull request by akhilome**
#### What does this PR do?

Implements favorite/unfavorite article functionality

#### Description of Task to be completed?

- add favorites model
- modify associated models to include favorite article feature
- modify article controller to include an article's fav stats
- add favorite article controller
- add favorite handler routes for favoriting and unfavoriting articles
- write tests for functionality
- document new API endpoints

#### How should this be manually tested?

- clone the repo, `cd` into it, and run `npm install` to install the project's dependencies,
- rename the `env.sample` file to `.env` and fill out the required environmental variables,
- start the local server by running `npm start`, then hit the create an account using the singup endpoint, per the documentation,
- After signing up, create a new article, by hitting the `POST /api/articles` endpoint.
- You can then favorite the article by hitting the `POST /api/articles/:slug/favorite` endpoint
- You can also unfavorite the article by hitting the `DELETE /api/articles/:slug/favorite` endpoint

#### Any background context you want to provide?

N/A

#### What are the relevant pivotal tracker stories?

[#162170716](https://www.pivotaltracker.com/story/show/162170716)

#### Screenshots (if appropriate)

screen shot 2018-12-09 at 6 05 12 pm

screen shot 2018-12-09 at 9 05 08 pm

#### Questions:

N/A

### CODE REVIEW

**Review by frostyblok — APPROVED** [review]

File: src/controllers/favoriteArticleController.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  static async favoriteArticle(request, response) {
+    const { slug } = request.params;
+    const userId = request.userData.payload.id;
+    try {
+      const articleId = (await ArticleModel.queryForArticle(slug)).id;
+      await FavoriteModelHelper.favoriteArticle(articleId, userId);
+
+      response.redirect(`/api/articles/${slug}`);
+    } catch (error) {
+      response.status(500).json({
```
Comment (frostyblok) [inline src/controllers/favoriteArticleController.js]: HI Kizito, I like your implementations and I have learned a lot, but I would like you to stick with the team's error response from the helper's method [here](https://github.com/andela/balder-ah-backend/blob/staging/src/helpers/index.js).

File: src/controllers/favoriteArticleController.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+      const articleId = (await ArticleModel.queryForArticle(slug)).id;
+      await FavoriteModelHelper.unfavoriteArticle(articleId, userId);
+      response.redirect(`/api/articles/${slug}`);
+    } catch (error) {
+      response.status(500).json({
+        message: 'Fail',
+        error: error.message
+      });
+    }
+  }
```
Comment (okoroemeka) [inline src/controllers/favoriteArticleController.js:44]: Nice work at @akhilome. Please, can you give me an overview of how this code works?

File: src/db/migrations/20181126141320-create-user.js [hunk 3]
```diff
@@ -40,6 +40,14 @@ export default {
       type: Sequelize.STRING,
       allowNull: true
     },
+    passwordExpiresIn: {
```
Comment (okoroemeka) [inline src/db/migrations/20181126141320-create-user.js]: why are you making use of passwordExpiresIn and token?

**Comment by okoroemeka** [comment]
@akhilome why do you still have `WIP` in your PR's `How should this be manually tested?`

**Review by okoroemeka — APPROVED** [review]

**Comment by akhilome** [comment]
> @akhilome why do you still have `WIP` in your PR's `How should this be manually tested?`

My bad, @okoroemeka, thanks for catching that, this has been fixed now.

File: src/controllers/favoriteArticleController.js (hunk 1 above)
Comment (akhilome) [reply src/controllers/favoriteArticleController.js]: Thanks for the feedback, @frostyblok, will work on implementing this in a few.

File: src/controllers/article.js [hunk 4]
```diff
@@ -50,13 +51,21 @@ class ArticleController {
    */
   static async getAllArticles(request, response) {
     try {
-      const allArticles = await ArticleModel.getAllArticle();
-      if (!allArticles) {
+      let allArticles = await ArticleModel.getAllArticle();
+
+      if (!allArticles || !allArticles.length) {
```
Comment (marcdomain) [inline src/controllers/article.js]: What difference would it make if you used only **!allArticles.length** here

File: src/db/migrations/20181126141320-create-user.js (hunk 3 above)
Comment (akhilome) [reply src/db/migrations/20181126141320-create-user.js]: Per our discussion, this has been fixed. 👍

**Review by marcdomain — CHANGES_REQUESTED** [review]
Kindly attend to the comment I dropped for you.

File: src/controllers/article.js (hunk 4 above)
Comment (akhilome) [reply src/controllers/article.js]: 🤔 

Now that I think about it, none, whatsoever. 

Thanks for point this to me, @marcdomain.

**Review by marcdomain — APPROVED** [review]
Good job @akhilome the comment was properly attended to, and you raised the test coverage by 0.1%

**Comment by Johnsonojo** [comment]
Nice work @akhilome. I like the way you structured your implementation and the fact that you increased the overall coverage by 0.1%. Good job!

**Review by Johnsonojo — COMMENTED** [review]
Nice work @akhilome. I like the way you structured your implementation and the fact that you increased the overall coverage by 0.1%. Good job!

File: src/helpers/articles.js [hunk 5]
```diff
@@ -8,9 +8,9 @@ const { Article } = Models;
  */
 class ArticleModel {
   /**
-   * @description - This method is responsible for quering the database for an article
-   * @static
-   * @returns {object} - object representing response message
+   *
```
Comment (omokaro-faith) [inline src/helpers/articles.js]: @akhilome Why did you remove the description?

Comment (akhilome) [reply src/helpers/articles.js]: @omokaro-faith, I removed the description for this method (`createArticle`), because it was misleading. What I, however, failed to do was put in an appropriate one - which I'd get done within the hour.
